src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1784,52 +1778,100 @@
+                  ;; them.  An external insertion of a fact that matches
+                  ;; a negation or accumulator condition can cause logical retractions.
+                  (doseq [[alpha-roots fact-group] (get-alphas-fn facts)
+                          root alpha-roots]
+                    (alpha-activate root fact-group transient-memory transport transient-listener))
+                  (external-retract-loop get-alphas-fn transient-memory transport transient-listener)))
+
+              :retraction
+              (do
+                (l/retract-facts! transient-listener facts)
+
+                (binding [*pending-external-retractions* (atom facts)]
+                  (external-retract-loop get-alphas-fn transient-memory transport transient-listener)))))
+
+          (fire-rules* rulebase
+                       (:production-nodes rulebase)
+                       transient-memory
+                       transport
+                       transient-listener
+                       get-alphas-fn
+                       (uc/get-ordered-update-cache)))
+
+        #?(:cljs (throw (ex-info "Using the option for insertions and retractions to cancel each other out is not supported in ClojureScript"
+                                 {:session session :opts opts}))
+
+           :clj (do (let [insertions (into []
+                                           (comp (filter (fn [pending-op]
+                                                           (= (:type pending-op)
+                                                              :insertion)))
+                                                 (map :facts)
+                                                 cat)
+                                           pending-operations)
+                          retractions (into []
+                                            (comp (filter (fn [pending-op]
+                                                            (= (:type pending-op)
+                                                               :retraction)))
+                                                  (map :facts)
+                                                  cat)
+                                            pending-operations)
+                          update-cache (ca/get-cancelling-update-cache)]
+
+                      (binding [*current-session* {:rulebase rulebase
+                                                   :transient-memory transient-memory
+                                                   :transport transport
+                                                   :insertions (atom 0)
+                                                   :get-alphas-fn get-alphas-fn
+                                                   :pending-updates update-cache
+                                                   :listener transient-listener}]
+
+                        (doseq [[alpha-roots fact-group] (get-alphas-fn retractions)
```

Comment:
The alpha-retract will have a chain effect of removing partial join results from beta memory across the network. Those memory updates will not have visibility into the partial join results that are subsequently added by the insertions done here after retractions. 

I'm not talking about the eventual RHS inserts or retracts that may happen at the terminal nodes of the network.